src/l1_fetcher.rs

```diff
@@ -124,122 +123,108 @@ impl L1Fetcher {
+        let mut disable_polling = self.config.disable_polling;
+        if end_block.is_some() {
+            disable_polling = false;
+        }
 
-                    if let Some(current_block) = tx.block_number {
-                        let current_block = current_block.as_u64();
-                        if last_block < current_block {
-                            let mut metrics = metrics.lock().await;
-                            metrics.l1_blocks_processed += current_block - last_block;
-                            last_block = current_block;
-                        }
-                    }
+        // Split L1 block processing into three async steps:
+        // - BlockCommit event filter (main).
+        // - Referred L1 block fetch (tx).
+        // - Calldata parsing (parse).
+        let tx_handle =
+            self.spawn_tx_handle(hash_rx, calldata_tx, current_l1_block_number.as_u64());
+        let parse_handle = self.spawn_parse_handle(calldata_rx, sink)?;
+        let main_handle = self.spawn_main_handle(
+            hash_tx,
+            shutdown_rx,
+            current_l1_block_number,
+            end_block_number,
+            disable_polling,
+        )?;
 
-                    calldata_tx.send(tx.input).await.unwrap();
-                }
-            }
-        });
+        tx_handle.await?;
+        parse_handle.await?;
+        main_handle.await?;
 
-        let parse_handle = tokio::spawn({
-            let metrics = metrics.clone();
-            async move {
-                while let Some(calldata) = calldata_rx.recv().await {
-                    let blocks = match parse_calldata(&function, &calldata) {
-                        Ok(blks) => blks,
-                        Err(e) => {
-                            tracing::error!("failed to parse calldata: {e}");
-                            continue;
-                        }
-                    };
+        self.metrics.lock().await.print();
 
-                    for blk in blocks {
-                        // NOTE: Let's see if we want to increment this in batches, instead of each block individually.
-                        let mut metrics = metrics.lock().await;
-                        metrics.l2_blocks_processed += 1;
-                        metrics.latest_l2_block_nbr = blk.block_number;
-                        sink.send(blk).await.unwrap();
-                    }
-                }
-            }
-        });
+        Ok(())
+    }
 
-        let (shutdown_tx, mut shutdown_rx) = oneshot::channel();
-        let main_handle = tokio::spawn({
-            let provider_clone = self.provider.clone();
-            let snapshot_clone = self.snapshot.clone();
-            let metrics = metrics.clone();
-            let mut disable_polling = self.config.disable_polling;
-            let end_block = self
-                .config
-                .block_count
-                .map(|count| U64::from(self.config.start_block + count));
+    fn spawn_main_handle(
```

Comment:
```suggestion
    fn spawn_main_handler(
```
Minor nit: Maybe add `r` suffix to these function names? In this context I think there is the semantic difference between `handle` (the object) vs. `handler` (the action) :slightly_smiling_face: